src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -7903,56 +7918,261 @@ def apply(
                 "Snowpark pandas apply API doesn't yet support DataFrame or Series in 'args' or 'kwargs' of 'func'"
             )
 
-        # get input types of all data columns from the dataframe directly
-        input_types = self._modin_frame.get_snowflake_type(
-            self._modin_frame.data_column_snowflake_quoted_identifiers
-        )
+        if axis == 0:
```

Comment:
![Screenshot 2024-09-06 at 4 09 07 PM](https://github.com/user-attachments/assets/827d0c58-237a-4942-b2ef-06d60fad7dd9)
why not just reuse groupby.apply to implement this?

Reply:
The basic approach is to use groupby apply, however it is more nuanced, check out the last part of the doc (https://docs.google.com/document/d/1P2HsPYKXt_RQ-nevtGohwPercffHKNc8R_aDmHHzQnE/edit) which basically shows how to do this.  The fake_group index part can be avoided internally as there is a way to force grouping the entire index which is what we want in this case.  Note that I'm not able to reproduce your example above.

![Screenshot 2024-09-09 at 11 00 32 AM](https://github.com/user-attachments/assets/9f3f7e73-4c87-4de2-bebe-067ea4ff931f)

Also the behavior is slightly different if the apply_func maps to series vs scalar or combinations across columns.  The final result could be a dataframe, series or scalar.  The apply function is applied to each column separately and then results and joined or concatenated together as appropriate.  There's more book keeping necessary to handle all of these cases.